openrarity/collection/collection.py

```diff
@@ -0,0 +1,337 @@
+
+
+class AttributesStatistics(TypedDict):
+    string: list[AttributeStatistic]
+    number: list[AttributeStatistic]
+    date: list[AttributeStatistic]
+
+
+class TokenCollection:
+    """ """
+
+    def __init__(
+        self,
+        token_type: Literal["non-fungible", "semi-fungible"],
+        tokens: dict[TokenId, RawToken],
+    ):
+        self._token_type = token_type
+        self._input_checksum: str = self._hash_data(cast(JsonEncodable, tokens))
+        self._ranks_checksum: str | None = None
+        self._token_supply: int | dict[str | int, int]
+        (self._token_supply, self._tokens) = validate_tokens(token_type, tokens)
+
+        # Derived data
+        self._vertical_attribute_data: list[ValidatedTokenAttribute] = []
+        self._attribute_statistics: AttributesStatistics = {
+            "string": [],
+            "number": [],
+            "date": [],
+        }
+        self._entropy: float | None = None
+        self._token_statistics: list[TokenStatistic] = []
+        self._string_types: list[ValidatedTokenAttribute] = []
+        self._number_types: list[ValidatedTokenAttribute] = []
+        self._date_types: list[ValidatedTokenAttribute] = []
+
+        # Output data
+        self._ranks: list[RankedToken] = []
+
+    def __repr__(self) -> str:
+        return f"Collection({self._token_type})"
+
+    @property
+    def tokens(self) -> dict[TokenId, RawToken]:
+        return self._tokens
+
+    @property
+    def total_supply(self) -> int:
+        # SemiFungible needs to sum the supply of each token
+        if isinstance(self._token_supply, dict):
+            return sum(self._token_supply.values())
+
+        return self._token_supply
+
+    @property
+    def attribute_statistics(self) -> AttributesStatistics:
+        if not self._attribute_statistics:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._attribute_statistics
+
+    @property
+    def entropy(self) -> float | None:
+        return self._entropy
+
+    @property
+    def token_statistics(self) -> list[TokenStatistic]:
+        if not self._token_statistics:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._token_statistics
+
+    @property
+    def ranks(self) -> list[RankedToken]:
+        if not self._ranks:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._ranks
+
+    @property
+    def checksum(self) -> str:
+        if not self._ranks_checksum:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._input_checksum + self._ranks_checksum
+
+    @overload
+    def rank_collection(  # type: ignore
```

Comment:
need better documentation in methods , otherwsie not clear why those overloads exists and why we need them